src/Orleans.Pipeline.Client/OrleansPipe.cs

```diff
@@ -28,97 +28,99 @@ internal class OrleansPipe<TToServer, TFromServer>(
     public ChannelReader<TFromServer> Reader => _reader.Reader;
     public ChannelWriter<TToServer> Writer => _writer.Writer;
 
-    public Task Start(CancellationToken token)
+    public async Task Start(CancellationToken token)
     {
         //get grain
         _grain = clusterClient.GetGrain<IOrleansPipeGrain<TToServer, TFromServer>>(id);
 
         //create task for writer invocation
-        _writerTask = RunWriter();
+        _writerTask = RunWriter(token);
 
         //create observer for reader
-        _observerTask = RunObserver();
+        _observerTask = RunObserver(token);
         _observer = clusterClient.CreateObjectReference<IOrleansPipeObserver<TFromServer>>(this);
-        return _grain.Subscribe(_observer);
+        await _grain.Subscribe(_observer);
     }
+
     public async Task Stop(CancellationToken token)
     {
-        //finish writer channel
-        _writer.Writer.Complete();
-        _writerStoppingTokenSource.Cancel();
-        if (_writerTask is not null)
+        try
         {
-            await _writerTask;
-        }
+            // Cancel any ongoing operations
+            _writer.Writer.Complete();
+            await (_writerTask ?? Task.CompletedTask).ConfigureAwait(ConfigureAwaitOptions.SuppressThrowing);
+            _writerStoppingTokenSource.Cancel();
```

Comment:
If we await the task ending bevor we cancel issnt this never ending, or does this ConfiguraAwaitOption not await the Task itself, but the configurationAction?

Reply:
@BernhardPollerspoeck we call `Complete()` on the `_writer.Writer` before that, this is propagated to the `_writer.Reader` which breaks out of the loop. `ConfigureAwaitOptions.SuppressThrowing` is just to ensure any potential exception which is not an `OperationCancelledException` (since that is caught by the writer task) and might be thrown won't shortcut the stopping process and ensure that observer unsubscribes and the object ref is removed (to avoid a memory leak)

Reply:
Technically there isn't even a need for the `_writerStoppingTokenSource`